CRM/Event/Form/EventFormTrait.php

```diff
@@ -92,4 +100,22 @@ public function getParticipantValue(string $fieldName) {
     return NULL;
   }
 
+  /**
+   * Get the number of available spaces in the given event.
+   *
+   * @internal this is a transitional function - eventually it will be removed
+   * and the api will handle it.
+   *
+   * @return int
+   *
+   * @throws \CRM_Core_Exception
+   */
+  protected function getAvailableSpaces(): int {
+    $availableSpaces = CRM_Event_BAO_Participant::eventFull($this->getEventID(),
+      TRUE,
+      $this->getEventValue('has_waitlist')
+    );
+    return is_numeric($availableSpaces) ? (int) $availableSpaces : 0;
```

Comment:
Regarding https://lab.civicrm.org/dev/core/-/issues/4918, I think the problem is this function is trying to convert something that is inherently a mixed value into a single integer. It doesn't take into account that the available spaces could be infinite (which is not an integer, and is what NULL means here - i.e. not the same as 0).

It's true that it feels weird to have a function that sometimes returns a display string and sometimes a number, but the underlying thing is that it is truly a mixed value, of some kind.

One option is that if it's NULL, return INT_MAX. But that feels kludgy, and what if you're hosting a video party for the World Cup Final...

Reply:
Right but this function is ALWAYS returning 0 available spaces left when you leave the maximum number of participants field blank, because CRM_Event_BAO_Participant::eventFull() returns null in that case, and then this function converts null to 0 instead of converting it to "infinity".

The World Cup reference was partly a joke, that if instead of returning infinity, because that's not an integer, you return php's highest integer, but for such an event that might not be big enough.

When I tried that though I then started seeing some problems with the confirmation page and haven't looked closer yet.

Reply:
Every event has a limit to the number of participants, but in some cases you may not want to cap the number of people who register for an event.

An example is where you're using an event to track volunteer applications. You may only need 20 volunteers, but you want to select the 20 volunteers that have the best skillsets for the roles. So you would allow everyone to register their interest to volunteer for the event, and then manually identify the people who you will ask to volunteer.

Looking at the `eventFull` method, it returns `NULL` if there is no participant limit. ([link](https://github.com/eileenmcnaughton/civicrm-core/blob/b0af8a91d55a6cddd71ca0582fd2f0fd194c7777/CRM/Event/BAO/Participant.php#L429)) which seems like a fair result rather than `0`.

Reply:
Just noting that the regression under discussion is now partially fixed and the remaining part is in  https://lab.civicrm.org/dev/core/-/issues/4942